apache/beam
# [BEAM-11105] Add timestamp observing watermark estimation

**Summary of Changes**

This pr is a continuation of building watermark support in the Go Sdk. It builds on the work done in https://github.com/apache/beam/pull/17374 and introduces the ability for watermark estimators to be updated based on the timestamp of emitted elements. It includes the following changes:

- Execution (emit/emit.go, exec/pardo.go, sdf/sdf.go)
- Code generation (generate/generate.go, **.shims.go, optimized/emitters.tmpl, optimized/emitters.go)
- Concrete implementation of a timestamp observing estimator (sdf/watermark_estimator.go)

This is built according to the design specified here - https://docs.google.com/document/d/1DqCYJ-J1YGNelCRIcN5v6BQsZxJB2l5uWBmN4ti--Ew/edit?usp=sharing

**Next Steps**

The following work remains to be implemented after this is completed

- Manual estimators (PR in progress - https://github.com/apache/beam/pull/17475)
- Doc changes

**Manual Testing**

On top of all the unit tests introduced, I hacked the existing textio sdf doFn and added the following function:

```
func (fn *readSdfFn) CreateWatermarkEstimator(t time.Time) *sdf.TimestampObservingWatermarkEstimator {
	return &sdf.TimestampObservingWatermarkEstimator{State: t}
}

func (fn *readSdfFn) InitialWatermarkEstimatorState(_ beam.EventTime, _ offsetrange.Restriction, _ string, _ int64) time.Time {
	return time.Now().AddDate(-1, 0, 0)
}

func (fn *readSdfFn) WatermarkEstimatorState(e *sdf.TimestampObservingWatermarkEstimator) time.Time {
	return e.State
}
```

and emitted elements 10 years in the future from processElement. Then I ran wordCount (with the textio sdf version) on dataflow so that I could observe the output watermark was getting correctly updated to the current wall clock time + 1 month:

<img width="437" alt="image" src="https://user-images.githubusercontent.com/42773683/165524201-8a16ad7f-4ca0-4d46-9b8d-37dcb9d8eb5c.png">

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:


See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.
